From 42f4e0b77c373c0632197fca79c5094c197abaf2 Mon Sep 17 00:00:00 2001
From: Mingliang Liu <liuml07@apache.org>
Date: Thu, 31 Aug 2017 22:36:56 -0700
Subject: [PATCH 2629/2848] HDFS-12363. Possible NPE in
 BlockManager$StorageInfoDefragmenter#scanAndCompactStorages.
 Contributed by Xiao Chen

(cherry picked from commit 1fbb662c7092d08a540acff7e92715693412e486)

Change-Id: I7ca0da468041e9c705017c85e39c70804bb8f49b
---
 .../hdfs/server/blockmanagement/BlockManager.java  |    8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
index f49f500..30adf38 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java
@@ -3951,8 +3951,12 @@ private void scanAndCompactStorages() throws InterruptedException {
         for (int i = 0; i < datanodesAndStorages.size(); i += 2) {
           namesystem.writeLock();
           try {
-            DatanodeStorageInfo storage = datanodeManager.
-                getDatanode(datanodesAndStorages.get(i)).
+            final DatanodeDescriptor dn = datanodeManager.
+                getDatanode(datanodesAndStorages.get(i));
+            if (dn == null) {
+              continue;
+            }
+            final DatanodeStorageInfo storage = dn.
                 getStorageInfo(datanodesAndStorages.get(i + 1));
             if (storage != null) {
               boolean aborted =
-- 
1.7.9.5

